pkg/proc/native/proc_freebsd.go

```diff
@@ -0,0 +1,366 @@
+}
+
+// Attach to an existing process with the given PID. Once attached, if
+// the DWARF information cannot be found in the binary, Delve will look
+// for external debug files in the directories passed in.
+func Attach(pid int, debugInfoDirs []string) (*Process, error) {
+	dbp := New(pid)
+	dbp.common = proc.NewCommonProcess(true)
+
+	var err error
+	dbp.execPtraceFunc(func() { err = PtraceAttach(dbp.pid) })
+	if err != nil {
+		return nil, err
+	}
+	_, _, err = dbp.wait(dbp.pid, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	err = dbp.initialize(findExecutable("", dbp.pid), debugInfoDirs)
+	if err != nil {
+		dbp.Detach(false)
+		return nil, err
+	}
+	return dbp, nil
+}
+
+func initialize(dbp *Process) error {
+	comm, _ := C.find_command_name(C.int(dbp.pid))
+	defer C.free(unsafe.Pointer(comm))
+	comm_str := C.GoString(comm)
+	dbp.os.comm = strings.Replace(string(comm_str), "%", "%%", -1)
+	return nil
+}
+
+// kill kills the target process.
+func (dbp *Process) kill() (err error) {
+	if dbp.exited {
+		return nil
+	}
+	dbp.execPtraceFunc(func() { err = PtraceCont(dbp.pid, int(sys.SIGKILL)) })
+	if err != nil {
+		return err
+	}
+	if _, _, err = dbp.wait(dbp.pid, 0); err != nil {
+		return err
+	}
+	dbp.postExit()
+	return nil
+}
+
+// Used by RequestManualStop
+func (dbp *Process) requestManualStop() (err error) {
+	return sys.Kill(dbp.pid, sys.SIGTRAP)
+}
+
+// Attach to a newly created thread, and store that thread in our list of
+// known threads.
+func (dbp *Process) addThread(tid int, attach bool) (*Thread, error) {
+	if thread, ok := dbp.threads[tid]; ok {
+		return thread, nil
+	}
+
+	var err error
+	dbp.execPtraceFunc(func() { err = sys.PtraceLwpEvents(dbp.pid, 1) })
+	if err == syscall.ESRCH {
+		if _, _, err = dbp.waitFast(dbp.pid); err != nil {
+			return nil, fmt.Errorf("error while waiting after adding process: %d %s", dbp.pid, err)
+		}
+	}
+
+	dbp.threads[tid] = &Thread{
+		ID:  tid,
+		dbp: dbp,
+		os:  new(OSSpecificDetails),
+	}
+
+	if dbp.currentThread == nil {
+		dbp.SwitchThread(tid)
+	}
+
+	return dbp.threads[tid], nil
+}
+
+// Used by initialize
+func (dbp *Process) updateThreadList() error {
+	var tids []int32
+	dbp.execPtraceFunc(func() { tids = PtraceGetLwpList(dbp.pid) })
+	for _, tid := range tids {
+		if _, err := dbp.addThread(int(tid), false); err != nil {
+			return err
+		}
+	}
+	dbp.os.tid = int(tids[0])
+	return nil
+}
+
+// Used by Attach
+func findExecutable(path string, pid int) string {
+	if path == "" {
+		cstr := C.find_executable(C.int(pid))
+		defer C.free(unsafe.Pointer(cstr))
+		path = C.GoString(cstr)
+	}
+	return path
+}
+
+func (dbp *Process) trapWait(pid int) (*Thread, error) {
+	return dbp.trapWaitInternal(pid, false)
+}
+
+// Used by stop and trapWait
+func (dbp *Process) trapWaitInternal(pid int, halt bool) (*Thread, error) {
+	for {
+		wpid, status, err := dbp.wait(pid, 0)
+		if err != nil {
+			return nil, fmt.Errorf("wait err %s %d", err, pid)
+		}
+		if status.Killed() {
```

Comment:
I wonder if this has anything to do with the failing concurrent tests, since we're effectively ignoring `status.Killed`, yes we send `SIGKILL` to the process in `process.kill()`. I have a freebsd VM I can test this theory out on tomorrow.